theano/tensor/nnet/nnet.py

```diff
@@ -786,16 +934,102 @@ def local_logsoftmax_grad(node):
                        subtensor.AdvancedIncSubtensor))):
         # get parameters from unoptimized op
         sm = node.inputs[0].owner.inputs[1]
+        axis = sm.owner.op.axis
         # sm_input = node.inputs[1].owner.inputs[0]
         grads = node.inputs[0].owner.inputs[0]
-        if grads.broadcastable[1] and not sm.broadcastable[1]:
-            grads = tensor.alloc(grads, grads.shape[0], sm.shape[1])
-        ret = grads - tensor.sum(grads, axis=1, keepdims=True) * sm
+        if grads.broadcastable[axis] and not sm.broadcastable[axis]:
+            grads = tensor.alloc(grads, grads.shape[0], sm.shape[axis])
+        ret = grads - tensor.sum(grads, axis=axis, keepdims=True) * sm
         ret.tag.values_eq_approx = values_eq_approx_remove_nan
         copy_stack_trace(node.outputs[0], ret)
         return [ret]
 
 
+@opt.register_specialize('fast_compile')
+@gof.local_optimizer([SoftmaxGrad])
+def local_indexing_logsoftmax_grad(node):
+    # One case is supported:
+    # 1. AdvancedIncSubtensor(
+    #           zeros_like(softmax(x)),
+    #           -out_grad / AdvancedSubtensor(softmax(x), arange(y.shape[0]), y),
```

Comment:
actually, it does not have to be an `AdvancedSubtensor`, nor `arange(y.shape[0])` as first argument.
I think you can just represent it as `SubtensorOp(softmax(x), *idx)`